server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -680,72 +689,157 @@ public String toString() {
                                     requestSeqIdTracker.markSeqNoAsCompleted(requestSeqId);
-                                }
+                                    sendFileChunks(listener);
+                                },
+                                e -> listener.onFailure(handleErrorOnSendFiles(chunk.md, e))
                             )));
-                    position += content.length();
+                } catch (Exception e) {
+                    listener.onFailure(e);
+                    return;
                 }
-            } catch (Exception e) {
-                error.compareAndSet(null, Tuple.tuple(md, e));
-                break;
             }
         }
-        // When we terminate exceptionally, we don't wait for the outstanding requests as we don't use their results anyway.
-        // This allows us to end quickly and eliminate the complexity of handling requestSeqIds in case of error.
-        if (error.get() == null) {
-            cancellableThreads.execute(() -> requestSeqIdTracker.waitForOpsToComplete(requestSeqIdTracker.getMaxSeqNo()));
+
+        synchronized FileChunk readChunk(final byte[] buffer) throws Exception {
+            cancellableThreads.checkForCancel();
+            if (readerClosed) {
+                throw new IllegalStateException("chunk reader was closed");
+            }
+            try {
+                if (currentInput == null) {
+                    if (files.isEmpty()) {
+                        return null;
+                    }
+                    md = files.remove(0);
+                    position = 0;
+                    final IndexInput indexInput = store.directory().openInput(md.name(), IOContext.READONCE);
+                    currentInput = new BufferedInputStream(new InputStreamIndexInput(indexInput, md.length()), chunkSizeInBytes) {
+                        @Override
+                        public void close() throws IOException {
+                            indexInput.close(); //InputStreamIndexInput's close is noop
+                        }
+                    };
+                }
+                final int bytesRead = currentInput.read(buffer, 0, buffer.length);
+                final boolean lastChunk = position + bytesRead == md.length();
+                if (bytesRead == -1) {
+                    IOUtils.close(currentInput, () -> currentInput = null);
+                    return readChunk(buffer);
+                }
+                final FileChunk chunk = new FileChunk(md, new BytesArray(buffer, 0, bytesRead), position, lastChunk);
+                position += bytesRead;
+                return chunk;
+            } catch (Exception e) {
+                throw handleErrorOnSendFiles(md, e);
+            }
         }
-        if (error.get() != null) {
-            handleErrorOnSendFiles(store, error.get().v1(), error.get().v2());
+
+        @Override
+        public synchronized void close() throws IOException {
```

Comment:
@dnhatn seems the test failure is related to this change https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-1/9612/testReport/junit/org.elasticsearch.repositories.fs/FsBlobStoreRepositoryIT/testSnapshotAndRestore/  and we're dead-locking here

Reply:
@original-brownbear Thanks for looking. I push https://github.com/elastic/elasticsearch/pull/39769/commits/e9e3584b249df521cfd969540c05b44cd3e0a2ad.